### PR TITLE
Enable mono-state-based telemetry for crashing on CI

### PR DESCRIPTION
**Pull request by alexanderkyte**
Everything seemed to be working before on CI before. Let's add the jenkins hooks for the crash files with this PR too.

### CODE REVIEW

File: mono/metadata/object-internals.h [hunk 1]
```diff
@@ -631,7 +631,7 @@ typedef struct {
 	gpointer (*create_delegate_trampoline) (MonoDomain *domain, MonoClass *klass);
 	gpointer (*interp_get_remoting_invoke) (gpointer imethod, MonoError *error);
 	GHashTable *(*get_weak_field_indexes) (MonoImage *image);
-#ifdef TARGET_OSX
+#ifndef HOST_WIN32
```
Comment (jaykrell) [inline mono/metadata/object-internals.h]: Should we just have the same data structures for all platforms where easy? Even if we don't use the  fields?

File: mono/mini/mini-exceptions.c [hunk 2]
```diff
@@ -2990,12 +2990,11 @@ mono_handle_native_crash (const char *signal, void *ctx, MONO_SIG_HANDLER_INFO_T
 			if (!dump_for_merp) {
 #ifdef DISABLE_STRUCTURED_CRASH
```
Comment (jaykrell) [inline mono/mini/mini-exceptions.c]: Consider compile-if instead  of preprocessor-if. It makes for better coverage when building other  platforms. i.e. less easy to break what you don't  build. ifdef is overrated and overused.

Comment (alexanderkyte) [reply mono/mini/mini-exceptions.c]: I usually cppcheck and that does a good job of checking that the code behind all ifdefs are legal

**Comment by luhenry** [comment]
@alexanderkyte I added the `do-not-merge` label because of the `[WIP]` in the tittle, feel free to remove the label as soon as you deem it necessary.

File: mono/mini/mini-exceptions.c (hunk 2 above)
Comment (alexanderkyte) [reply mono/mini/mini-exceptions.c]: And this is partially why I'm doing what I do here with setting `leave = TRUE`, so the code below is enabled or disabled based on a compile-if.

**Comment by alexanderkyte** [comment]
```
make: Entering directory '/mnt/jenkins/workspace/test-mono-pull-request-wasm/sdks/wasm'
(cd test-suite && ~/.jsvu/sm test.js system-core)
/bin/bash: /home/builder/.jsvu/sm: No such file or directory
Makefile:160: recipe for target 'run-sm-system-core' failed``` Hmm

**Comment by alexanderkyte** [comment]
Failed lane: Linux x64 Interpreter:  expected fullaotmixed-regression

Failed lane: Linux x64 FullAOT+LLVM
```20:31:21 ***** MonoTests.System.Linq.ParallelEnumerableTests.TestCast
20:31:21 ***** MonoTests.System.Linq.ParallelEnumerableTests.TestFirst
20:31:21 ***** MonoTests.System.Linq.ParallelEnumerableTests.TestFirstOrDefault
20:31:21 ***** MonoTests.System.Linq.ParallelEnumerableTests.TestGroupBy
20:31:21 Stacktrace:
20:31:21 
20:31:21 /proc/self/maps:
20:31:21 5649d2e8c000-5649d3a01000 r-xp 00000000 08:11 32780187                   /mnt/jenkins/workspace/test-mono-pull-request-amd64-fullaot+llvm/mono/mini/mono-sgen
20:31:21 5649d3c00000-5649d3c20000 r--p 00b74000 08:11 32780187                   /mnt/jenkins/workspace/test-mono-pull-request-amd64-fullaot+llvm/mono/mini/mono-sgen
20:31:21 5649d3c20000-5649d3c27000 rw-p 00b94000 08:11 32780187                   /mnt/jenkins/workspace/test-mono-pull-request-amd64-fullaot+llvm/mono/mini/mono-sgen
20:31:21 5649d3c27000-5649d3c47000 rw-p 00000000 00:00 0 
20:31:21 5649d5867000-5649d6ece000 rw-p 00000000 00:00 0                          [heap]
20:31:21 7f7cb8000000-7f7cb8038000 rw-p 00000000 00:00 0 
20:31:21 7f7cb8038000-7f7cbc000000 ---p 00000000 00:00 0 ```

Failed lane: Linux ARMv7
Same above trace for sgen-new-threads-dont-join-stw-2.exe

Did I break sgen? Seems unlikely.

**Comment by lewurm** [comment]
@monojenkins build failed

**Comment by directhex** [comment]
Which lanes do you want mono_crash.*.json stored, and which directory does it go in?

**Comment by alexanderkyte** [comment]
@directhex it's dropped in the cwd of the runtime on a crash

**Comment by directhex** [comment]
@alexanderkyte added to all PR jobs, I think

**Comment by alexanderkyte** [comment]
@monojenkins build failed

File: mono/metadata/object-internals.h [hunk 3]
```diff
@@ -665,7 +663,7 @@ typedef struct {
 	gboolean (*mono_above_abort_threshold) (void);
 	void (*mono_clear_abort_threshold) (void);
 	void (*mono_reraise_exception) (MonoException *ex);
-#ifdef TARGET_OSX
+#ifndef HOST_WIN32
```
Comment (jaykrell) [inline mono/metadata/object-internals.h]: Try to remove ifdef?

File: mono/metadata/threads-types.h [hunk 4]
```diff
@@ -342,7 +342,7 @@ mono_set_thread_dump_dir(gchar* dir);
 MONO_COLD void
 mono_set_pending_exception_handle (MonoExceptionHandle exc);
 
-#ifdef TARGET_OSX
+#ifndef HOST_WIN32
```
Comment (jaykrell) [inline mono/metadata/threads-types.h]: Try to remove ifdef?

File: mono/mini/mini-posix.c [hunk 5]
```diff
@@ -1036,7 +1034,6 @@ dump_native_stacktrace (const char *signal, void *ctx)
 #endif
 			}
 
-#ifdef TARGET_OSX
```
Comment (jaykrell) [inline mono/mini/mini-posix.c]: This is the main part, right?

**Comment by lewurm** [comment]
I suggest to introduce a descriptive define instead of sprinkling `#ifndef HOST_WIN32` around everywhere

File: mono/metadata/object-internals.h (hunk 3 above)
Comment (vargaz) [reply mono/metadata/object-internals.h]: Yes, putting ifdefs in structures which are not perf critical is not very useful.

**Comment by jaykrell** [comment]
Or remove the ifdefs, at least most of them.

**Comment by jaykrell** [comment]
The Linux x64 FullAOT+LLVM failure is common.
https://github.com/mono/mono/issues/9554

**Comment by alexanderkyte** [comment]
@lewurm it would have to be defined at a really global place, like at configure.ac for it to cover such a wide array of files

File: mono/mini/mini-posix.c (hunk 5 above)
Comment (alexanderkyte) [reply mono/mini/mini-posix.c]: It's one of two external entry points into the system

File: winconfig.h [hunk 6]
```diff
@@ -83,6 +83,9 @@
 /* Disable support debug logging */
 /* #undef DISABLE_LOGGING */
 
+/* Disable runtime state dumping */
+#define DISABLE_CRASH_REPORTING
```
Comment (lewurm) [inline winconfig.h]:
```c
#define DISABLE_CRASH_REPORTING 1
```

**Comment by alexanderkyte** [comment]
@monojenkins build failed

**Comment by alexanderkyte** [comment]
@monojenkins build failed

**Comment by alexanderkyte** [comment]
https://github.com/mono/mono/pull/9755 fixes the failure we're seeing on CI here. Cherry-picked that commit.

**Comment by alexanderkyte** [comment]
The acceptance test lane looks like it's saying that I broke coreclr? 🤔

**Comment by alexanderkyte** [comment]
The acceptance test spurious failure is evidence of this working!
https://jenkins.mono-project.com/job/test-mono-pull-request-acceptance-tests/7058/Azure/processDownloadRequest/7058/acceptance-tests/mono_crash.15803.0.json

**Comment by alexanderkyte** [comment]
ios failure does not appear related:
```

make[1]: Leaving directory '/Users/builder/jenkins/workspace/test-mono-pull-request-amd64-osx-products-sdks-ios/sdks/ios'
make[1]: Entering directory '/Users/builder/jenkins/workspace/test-mono-pull-request-amd64-osx-products-sdks-ios/sdks/ios'
mono appbuilder.exe --target ios-sim64 --mono-sdkdir /Users/builder/jenkins/workspace/test-mono-pull-request-amd64-osx-products-sdks-ios/sdks/ios/../out --appdir /Users/builder/jenkins/workspace/test-mono-pull-request-amd64-osx-products-sdks-ios/sdks/ios/bin/ios-sim/test-System.Xml.Linq.app --runtimedir /Users/builder/jenkins/workspace/test-mono-pull-request-amd64-osx-products-sdks-ios/sdks/ios/runtime --builddir obj/ios-sim/test-System.Xml.Linq.app --sysroot /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS.sdk --signing-identity - --bundle-executable test-System.Xml.Linq --bundle-identifier com.xamarin.mono.ios.test-System.Xml.Linq --bundle-name test-System.Xml.Linq -r ../../mcs/class/lib/monotouch/mscorlib.dll -r ../../mcs/class/lib/monotouch/System.dll -r ../../mcs/class/lib/monotouch/System.Xml.dll -r ../../mcs/class/lib/monotouch/System.Core.dll -r ../../mcs/class/lib/monotouch/I18N.dll -r ../../mcs/class/lib/monotouch/I18N.West.dll -r ../../mcs/class/lib/monotouch/Mono.Simd.dll -r ../../mcs/class/lib/monotouch/Mono.Security.dll -r ../../mcs/class/lib/monotouch/System.Numerics.dll -r ../../mcs/class/lib/monotouch/System.Numerics.Vectors.dll -r ../../mcs/class/lib/monotouch/nunitlite.dll -r test-runner.exe -r ../../mcs/class/lib/monotouch/System.Xml.Linq.dll -r ../../mcs/class/lib/monotouch/tests/monotouch_System.Xml.Linq_test.dll
mkdir -p /Users/builder/jenkins/workspace/test-mono-pull-request-amd64-osx-products-sdks-ios/sdks/ios/bin/ios-sim/test-System.Xml.Linq.app
if test "x" != "x"; then cp -r  /Users/builder/jenkins/workspace/test-mono-pull-request-amd64-osx-products-sdks-ios/sdks/ios/bin/ios-sim/test-System.Xml.Linq.app/; fi
ninja -C obj/ios-sim/test-System.Xml.Linq.app -v
ninja: Entering directory `obj/ios-sim/test-System.Xml.Linq.app'
[1/22] if cmp -s System.dll /Users/builder/jenkins/workspace/test-mono-pull-request-amd64-osx-products-sdks-ios/sdks/ios/bin/ios-sim/test-System.Xml.Linq.app/System.dll ; then : ; else cp System.dll /Users/builder/jenkins/workspace/test-mono-pull-request-amd64-osx-products-sdks-ios/sdks/ios/bin/ios-sim/test-System.Xml.Linq.app/System.dll ; fi
[2/22] if cmp -s mscorlib.dll /Users/builder/jenkins/workspace/test-mono-pull-request-amd64-osx-products-sdks-ios/sdks/ios/bin/ios-sim/test-System.Xml.Linq.app/mscorlib.dll ; then : ; else cp mscorlib.dll /Users/builder/jenkins/workspace/test-mono-pull-request-amd64-osx-products-sdks-ios/sdks/ios/bin/ios-sim/test-System.Xml.Linq.app/mscorlib.dll ; fi
[3/22] if cmp -s System.Xml.dll /Users/builder/jenkins/workspace/test-mono-pull-request-amd64-osx-products-sdks-ios/sdks/ios/bin/ios-sim/test-System.Xml.Linq.app/System.Xml.dll ; then : ; else cp System.Xml.dll /Users/builder/jenkins/workspace/test-mono-pull-request-amd64-osx-products-sdks-ios/sdks/ios/bin/ios-sim/test-System.Xml.Linq.app/System.Xml.dll ; fi
[4/22] if cmp -s System.Core.dll /Users/builder/jenkins/workspace/test-mono-pull-request-amd64-osx-products-sdks-ios/sdks/ios/bin/ios-sim/test-System.Xml.Linq.app/System.Core.dll ; then : ; else cp System.Core.dll /Users/builder/jenkins/workspace/test-mono-pull-request-amd64-osx-products-sdks-ios/sdks/ios/bin/ios-sim/test-System.Xml.Linq.app/System.Core.dll ; fi
[5/22] if cmp -s I18N.dll /Users/builder/jenkins/workspace/test-mono-pull-request-amd64-osx-products-sdks-ios/sdks/ios/bin/ios-sim/test-System.Xml.Linq.app/I18N.dll ; then : ; else cp I18N.dll /Users/builder/jenkins/workspace/test-mono-pull-request-amd64-osx-products-sdks-ios/sdks/ios/bin/ios-sim/test-System.Xml.Linq.app/I18N.dll ; fi
[6/22] if cmp -s I18N.West.dll /Users/builder/jenkins/workspace/test-mono-pull-request-amd64-osx-products-sdks-ios/sdks/ios/bin/ios-sim/test-System.Xml.Linq.app/I18N.West.dll ; then : ; else cp I18N.West.dll /Users/builder/jenkins/workspace/test-mono-pull-request-amd64-osx-products-sdks-ios/sdks/ios/bin/ios-sim/test-System.Xml.Linq.app/I18N.West.dll ; fi
[7/22] if cmp -s Mono.Simd.dll /Users/builder/jenkins/workspace/test-mono-pull-request-amd64-osx-products-sdks-ios/sdks/ios/bin/ios-sim/test-System.Xml.Linq.app/Mono.Simd.dll ; then : ; else cp Mono.Simd.dll /Users/builder/jenkins/workspace/test-mono-pull-request-amd64-osx-products-sdks-ios/sdks/ios/bin/ios-sim/test-System.Xml.Linq.app/Mono.Simd.dll ; fi
[8/22] if cmp -s Mono.Security.dll /Users/builder/jenkins/workspace/test-mono-pull-request-amd64-osx-products-sdks-ios/sdks/ios/bin/ios-sim/test-System.Xml.Linq.app/Mono.Security.dll ; then : ; else cp Mono.Security.dll /Users/builder/jenkins/workspace/test-mono-pull-request-amd64-osx-products-sdks-ios/sdks/ios/bin/ios-sim/test-System.Xml.Linq.app/Mono.Security.dll ; fi
[9/22] if cmp -s System.Numerics.dll /Users/builder/jenkins/workspace/test-mono-pull-request-amd64-osx-products-sdks-ios/sdks/ios/bin/ios-sim/test-System.Xml.Linq.app/System.Numerics.dll ; then : ; else cp System.Numerics.dll /Users/builder/jenkins/workspace/test-mono-pull-request-amd64-osx-products-sdks-ios/sdks/ios/bin/ios-sim/test-System.Xml.Linq.app/System.Numerics.dll ; fi
[10/22] if cmp -s System.Numerics.Vectors.dll /Users/builder/jenkins/workspace/test-mono-pull-request-amd64-osx-products-sdks-ios/sdks/ios/bin/ios-sim/test-System.Xml.Linq.app/System.Numerics.Vectors.dll ; then : ; else cp System.Numerics.Vectors.dll /Users/builder/jenkins/workspace/test-mono-pull-request-amd64-osx-products-sdks-ios/sdks/ios/bin/ios-sim/test-System.Xml.Linq.app/System.Numerics.Vectors.dll ; fi
[11/22] mkdir -p /Users/builder/jenkins/workspace/test-mono-pull-request-amd64-osx-products-sdks-ios/sdks/ios/bin/ios-sim/test-System.Xml.Linq.app
[12/22] if cmp -s nunitlite.dll /Users/builder/jenkins/workspace/test-mono-pull-request-amd64-osx-products-sdks-ios/sdks/ios/bin/ios-sim/test-System.Xml.Linq.app/nunitlite.dll ; then : ; else cp nunitlite.dll /Users/builder/jenkins/workspace/test-mono-pull-request-amd64-osx-products-sdks-ios/sdks/ios/bin/ios-sim/test-System.Xml.Linq.app/nunitlite.dll ; fi
[13/22] if cmp -s test-runner.exe /Users/builder/jenkins/workspace/test-mono-pull-request-amd64-osx-products-sdks-ios/sdks/ios/bin/ios-sim/test-System.Xml.Linq.app/test-runner.exe ; then : ; else cp test-runner.exe /Users/builder/jenkins/workspace/test-mono-pull-request-amd64-osx-products-sdks-ios/sdks/ios/bin/ios-sim/test-System.Xml.Linq.app/test-runner.exe ; fi
[14/22] if cmp -s System.Xml.Linq.dll /Users/builder/jenkins/workspace/test-mono-pull-request-amd64-osx-products-sdks-ios/sdks/ios/bin/ios-sim/test-System.Xml.Linq.app/System.Xml.Linq.dll ; then : ; else cp System.Xml.Linq.dll /Users/builder/jenkins/workspace/test-mono-pull-request-amd64-osx-products-sdks-ios/sdks/ios/bin/ios-sim/test-System.Xml.Linq.app/System.Xml.Linq.dll ; fi
[15/22] if cmp -s monotouch_System.Xml.Linq_test.dll /Users/builder/jenkins/workspace/test-mono-pull-request-amd64-osx-products-sdks-ios/sdks/ios/bin/ios-sim/test-System.Xml.Linq.app/monotouch_System.Xml.Linq_test.dll ; then : ; else cp monotouch_System.Xml.Linq_test.dll /Users/builder/jenkins/workspace/test-mono-pull-request-amd64-osx-products-sdks-ios/sdks/ios/bin/ios-sim/test-System.Xml.Linq.app/monotouch_System.Xml.Linq_test.dll ; fi
[16/22] cp /Users/builder/jenkins/workspace/test-mono-pull-request-amd64-osx-products-sdks-ios/sdks/ios/runtime/runtime /Users/builder/jenkins/workspace/test-mono-pull-request-amd64-osx-products-sdks-ios/sdks/ios/bin/ios-sim/test-System.Xml.Linq.app/test-System.Xml.Linq
[17/22] if cmp -s config.json /Users/builder/jenkins/workspace/test-mono-pull-request-amd64-osx-products-sdks-ios/sdks/ios/bin/ios-sim/test-System.Xml.Linq.app/config.json ; then : ; else cp config.json /Users/builder/jenkins/workspace/test-mono-pull-request-amd64-osx-products-sdks-ios/sdks/ios/bin/ios-sim/test-System.Xml.Linq.app/config.json ; fi
[18/22] cp -r /Users/builder/jenkins/workspace/test-mono-pull-request-amd64-osx-products-sdks-ios/sdks/ios/runtime/Base.lproj /Users/builder/jenkins/workspace/test-mono-pull-request-amd64-osx-products-sdks-ios/sdks/ios/bin/ios-sim/test-System.Xml.Linq.app/Base.lproj
[19/22] if cmp -s /Users/builder/jenkins/workspace/test-mono-pull-request-amd64-osx-products-sdks-ios/sdks/ios/runtime/Entitlements.xcent Entitlements.xcent ; then : ; else cp /Users/builder/jenkins/workspace/test-mono-pull-request-amd64-osx-products-sdks-ios/sdks/ios/runtime/Entitlements.xcent Entitlements.xcent ; fi
[20/22] cp Info.plist Info.plist.binary; plutil -convert binary1 Info.plist.binary
[21/22] if cmp -s Info.plist.binary /Users/builder/jenkins/workspace/test-mono-pull-request-amd64-osx-products-sdks-ios/sdks/ios/bin/ios-sim/test-System.Xml.Linq.app/Info.plist ; then : ; else cp Info.plist.binary /Users/builder/jenkins/workspace/test-mono-pull-request-amd64-osx-products-sdks-ios/sdks/ios/bin/ios-sim/test-System.Xml.Linq.app/Info.plist ; fi
[22/22] codesign --force --sign - --timestamp=none /Users/builder/jenkins/workspace/test-mono-pull-request-amd64-osx-products-sdks-ios/sdks/ios/bin/ios-sim/test-System.Xml.Linq.app/test-System.Xml.Linq
make[1]: Leaving directory '/Users/builder/jenkins/workspace/test-mono-pull-request-amd64-osx-products-sdks-ios/sdks/ios'
make[1]: Entering directory '/Users/builder/jenkins/workspace/test-mono-pull-request-amd64-osx-products-sdks-ios/sdks/ios'
mono harness.exe --start-sim
Running: xcrun simctl list devices
xcrun simctl install xamarin.ios-sdk.sim bin/ios-sim/test-System.Xml.Linq.app
mono harness.exe --run-sim --logfile ios-sim-System.Xml.Linq.log --bundle-id com.xamarin.mono.ios.test-System.Xml.Linq --bundle-dir bin/ios-sim/test-System.Xml.Linq.app  test-runner.exe CONNSTR -exclude:MobileNotWorking,NotOnMac,NotWorking,ValueAdd,CAS,InetAccess,NotWorkingLinqInterpreter -labels monotouch_System.Xml.Linq_test.dll
App: com.xamarin.mono.ios.test-System.Xml.Linq
Running: xcrun simctl list devices
Running: xcrun simctl install xamarin.ios-sdk.sim bin/ios-sim/test-System.Xml.Linq.app
Running: xcrun simctl terminate xamarin.ios-sdk.sim com.xamarin.mono.ios.test-System.Xml.Linq
Running: xcrun simctl launch xamarin.ios-sdk.sim com.xamarin.mono.ios.test-System.Xml.Linq test-runner.exe tcp:localhost:63085 -exclude:MobileNotWorking,NotOnMac,NotWorking,ValueAdd,CAS,InetAccess,NotWorkingLinqInterpreter -labels monotouch_System.Xml.Linq_test.dll 
com.xamarin.mono.ios.test-System.Xml.Linq: 71023
NUnitLite 1.0.0 (.NET 4.5 Debug)
Copyright 2013, Charlie Poole

Runtime Environment -
    OS Version: Unix 17.4.0.0
  Mono Version: 4.0.50524.0

***** /Users/builder/Library/Developer/CoreSimulator/Devices/4DA9E16A-DB64-4A0A-87D0-3A91C69D4F58/data/Containers/Bundle/Application/1DD9D82C-A865-4C87-B4F2-01D9B64EF772/test-System.Xml.Linq.app/monotouch_System.Xml.Linq_test.dll
***** MonoTests.System.Xml.ExtensionsTest
***** MonoTests.System.Xml.ExtensionsTest.DocumentWithProcessingInstruction
make[1]: *** [Makefile:136: run-ios-sim-System.Xml.Linq] Error 1
make[1]: Leaving directory '/Users/builder/jenkins/workspace/test-mono-pull-request-amd64-osx-products-sdks-ios/sdks/ios'
make: *** [Makefile:150: run-ios-sim-all] Error 1
make: Leaving directory '/Users/builder/jenkins/workspace/test-mono-pull-request-amd64-osx-products-sdks-ios/sdks/ios'
/Users/builder/jenkins/workspace/test-mono-pull-request-amd64-osx-products-sdks-ios/scripts/ci/babysitter: Test suite terminated with code 2, and suite cannot report test case data. Halting.
```

**Comment by marek-safar** [comment]
@monojenkins backport 2018-06

**Comment by marek-safar** [comment]
@monojenkins backport 2018-08

**Comment by marek-safar** [comment]
@alexanderkyte could you please backport manually

**Pull request merged**